### PR TITLE
Handle healthcheck info missing

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
fix: #29741

### CODE REVIEW

File: pilot/pkg/controller/workloadentry/workloadentry_controller.go [hunk 1]
```diff
@@ -279,21 +283,31 @@ func (c *Controller) UpdateWorkloadEntryHealth(proxy *model.Proxy, event HealthE
...
-	// get previous status
-	cfg := c.store.Get(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace)
-	if cfg == nil {
-		log.Errorf("config was nil when getting WorkloadEntry %v for %v", entryName, proxy.ID)
-		return
-	}
+	condition := transformHealthEvent(event)
+	// backoff retry
+	// TODO(@hzxuzhonghu): handle race when reconnect happens and health status changed
+	for i := 0; i < maxRetries; i++ {
```
Comment (howardjohn) [inline pilot/pkg/controller/workloadentry/workloadentry_controller.go:289]: This is going to block the entire ADS stream for the proxy. If we are going to do this might make more sense to just return an error?

File: pilot/pkg/controller/workloadentry/workloadentry_controller.go [hunk 2]
```diff
@@ -279,21 +283,31 @@ func (c *Controller) UpdateWorkloadEntryHealth(proxy *model.Proxy, event HealthE
...
+		if i > 0 {
+			time.Sleep(waitTime << (i - 1))
```
Comment (howardjohn) [inline pilot/pkg/controller/workloadentry/workloadentry_controller.go:291]: nit: can we use decimal math instead of bit math? Its a lot easier to grok for most

File: pilot/pkg/controller/workloadentry/workloadentry_controller.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/controller/workloadentry/workloadentry_controller.go:289]: oh nevermind, I see you moved it to a goroutine

File: pkg/istio-agent/health/health_check.go [hunk 3]
```diff
@@ -121,18 +118,14 @@ func orDefault(val int32, def int32) int32 {
...
 		return
 	}
-	healthCheckLog.SetOutputLevel(log.DebugLevel)
-	healthCheckLog.Infof("starting health check for %T in %v", w.prober, w.config.InitialDelay)
 
+	healthCheckLog.Infof("starting health check for %T in %v", w.prober, w.config.InitialDelay)
 	// delay before starting probes.
 	time.Sleep(w.config.InitialDelay)
 
-	healthCheckLog.Debugf("health check initial delay complete")
```
Comment (howardjohn) [inline pkg/istio-agent/health/health_check.go:134]: nit: I think this was useful

File: pkg/istio-agent/health/health_check.go [hunk 4]
```diff
@@ -152,7 +145,6 @@ func (w *WorkloadHealthChecker) PerformApplicationHealthCheck(callback func(*Pro
 			if numSuccess == w.config.SuccessThresh && !lastStateHealthy {
 				healthCheckLog.Info("success threshold hit, marking as healthy")
 				callback(&ProbeEvent{Healthy: true})
-				numSuccess = 0
```
Comment (howardjohn) [inline pkg/istio-agent/health/health_check.go:155]: Why is this removed? This means `numSuccess == w.config.SuccessThresh` will happen exactly 1 one time, and going from pass -> fail -> pass will not trigger an update?

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: pkg/istio-agent/health/health_check.go (hunk 4 above)
Comment (hzxuzhonghu) [reply pkg/istio-agent/health/health_check.go:155]: nope, the other branch will set it to 0 when it is unhealthy

**Comment by hzxuzhonghu** [comment]
I donot think this is a perfect solution, and there is potential race. Take a look at #29741

**Comment by hzxuzhonghu** [comment]
Continue working on this instead of abusing xds.

File: pkg/istio-agent/health/health_check.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pkg/istio-agent/health/health_check.go:134]: there is one log in L124

File: pkg/istio-agent/health/health_check_test.go [hunk 5]
```diff
@@ -124,7 +124,7 @@ func TestWorkloadHealthChecker_PerformApplicationHealthCheck(t *testing.T) {
 			{Healthy: true},
 			{Healthy: false},
 		}
-		httpHealthStatuses := [4]bool{true, false, true, false}
+		httpHealthStatuses := [4]bool{true, false, false, true}
```
Comment (howardjohn) [inline pkg/istio-agent/health/health_check_test.go:127]: shouldn't this match `expectedHTTPEvents`?

File: pilot/pkg/controller/workloadentry/workloadentry_controller.go [hunk 6]
```diff
@@ -333,21 +346,42 @@ func (c *Controller) UpdateWorkloadEntryHealth(proxy *model.Proxy, event HealthE
...
-	// replace the updated status
-	wle := status.UpdateConfigCondition(*cfg, transformHealthEvent(event))
-	// update the status
-	_, err := c.store.UpdateStatus(wle)
-	if err != nil {
-		log.Errorf("error while updating WorkloadEntry status: %v for %v", err, proxy.ID)
+	// backoff retry
+	for i := 0; i < maxRetries; i++ {
+		if i > 0 {
+			time.Sleep(time.Duration(float64(waitTime.Nanoseconds()) * math.Pow(2, float64(i-1))))
```
Comment (howardjohn) [inline pilot/pkg/controller/workloadentry/workloadentry_controller.go:355]: we use github.com/cenkalti/backoff.NewExponentialBackOff() in other places if you prefer

File: pilot/pkg/controller/workloadentry/workloadentry_controller.go [hunk 7]
```diff
@@ -333,21 +346,42 @@ func (c *Controller) UpdateWorkloadEntryHealth(proxy *model.Proxy, event HealthE
...
+		}
+		// get previous status
+		cfg := c.store.Get(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace)
+		if cfg == nil {
+			log.Errorf("config was nil when getting WorkloadEntry %v for %v", entryName, proxy.ID)
```
Comment (howardjohn) [inline pilot/pkg/controller/workloadentry/workloadentry_controller.go:360]: I think the error message can be more useful. This wouldn't really help a user much. Maybe something like
`log.Errorf("failed to update health status for %v: WorkloadEntry %v not found", proxy.ID, entryName)`

File: pilot/pkg/controller/workloadentry/workloadentry_controller.go [hunk 8]
```diff
@@ -333,21 +346,42 @@ func (c *Controller) UpdateWorkloadEntryHealth(proxy *model.Proxy, event HealthE
...
+		if condition == nil {
+			return
+		}
+
+		// replace the updated status
+		wle := status.UpdateConfigCondition(*cfg, condition)
+		// update the status
+		_, err := c.store.UpdateStatus(wle)
+		if err != nil {
+			log.Errorf("error while updating WorkloadEntry status: %v for %v", err, proxy.ID)
```
Comment (howardjohn) [inline pilot/pkg/controller/workloadentry/workloadentry_controller.go:379]: make the error message more specific that its about health status, we have mutliple  kinds of status

File: pilot/pkg/controller/workloadentry/workloadentry_controller.go [hunk 9]
```diff
@@ -333,21 +346,42 @@ func (c *Controller) UpdateWorkloadEntryHealth(proxy *model.Proxy, event HealthE
...
 	}
 
-	// get previous status
-	cfg := c.store.Get(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace)
-	if cfg == nil {
-		log.Errorf("config was nil when getting WorkloadEntry %v for %v", entryName, proxy.ID)
-		return
-	}
+	condition := transformHealthEvent(event)
+	c.storeHealthStatus(makeProxyKey(proxy), condition)
```
Comment (howardjohn) [inline pilot/pkg/controller/workloadentry/workloadentry_controller.go:350]: Why do we need to store it in the controller globally and then fetch it later in the same function? Couldn't we just use `condition` directly on L368?

Comment (howardjohn) [reply pilot/pkg/controller/workloadentry/workloadentry_controller.go:350]: I guess maybe if a new event comes in, then we will start off a new goroutine, so we need to use the latest condition. However, I feel like we could approach this different so we don't spin off multiple goroutines all trying to write to the same WE and instead consolidate them?

File: pkg/istio-agent/health/health_check_test.go (hunk 5 above)
Comment (hzxuzhonghu) [reply pkg/istio-agent/health/health_check_test.go:127]: This test is confusing. Actually the 2 and 3 healthcheck result does not change, so the 3rd check does not result into status change and no callback at all.

File: pilot/pkg/controller/workloadentry/workloadentry_controller.go (hunk 6 above)
Comment (hzxuzhonghu) [reply pilot/pkg/controller/workloadentry/workloadentry_controller.go:355]: good to know

File: pilot/pkg/controller/workloadentry/workloadentry_controller.go (hunk 7 above)
Comment (hzxuzhonghu) [reply pilot/pkg/controller/workloadentry/workloadentry_controller.go:360]: make sense

File: pilot/pkg/controller/workloadentry/workloadentry_controller.go (hunk 9 above)
Comment (hzxuzhonghu) [reply pilot/pkg/controller/workloadentry/workloadentry_controller.go:350]: looking

File: pilot/pkg/controller/workloadentry/workloadentry_controller.go [hunk 10]
```diff
@@ -98,10 +111,19 @@ type Controller struct {
...
-	// maxConnectionAge is a duration that workload entry should be cleanedup if it does not reconnects.
+	// maxConnectionAge is a duration that workload entry should be cleaned up if it does not reconnects.
 	maxConnectionAge time.Duration
+
+	// latest health check status of the connected workload entries
+	// When the workload entry disconnects, the status will be removed.
+	// keyed by proxy network+ip
+	latestHealthEvent map[string]*HealthStatus
+
+	healthCondition cache.Queue
```
Comment (howardjohn) [inline pilot/pkg/controller/workloadentry/workloadentry_controller.go:122]: nit: can you add a comment

**Review by howardjohn — APPROVED** [review]
The fake k8s client can inject failures, may be useful to test the retry a bit

**Comment by hzxuzhonghu** [comment]
I will fix the nits
